### ISSUE TITLE
Feature Request: checking for Untrusted input

### ISSUE THREAD

**Issue by azu** [rhysd/actionlint#19]
[Keeping your GitHub Actions and workflows secure: Untrusted input | GitHub Security Lab](https://securitylab.github.com/research/github-actions-untrusted-input/) describe the Untrusted input on GitHub Actions.

In many case, using next values in `run:` or `script:` directry has vulnerbility.

    github.event.issue.title
    github.event.issue.body
    github.event.pull_request.title
    github.event.pull_request.body
    github.event.comment.body
    github.event.review.body
    github.event.review_comment.body
    github.event.pages.*.page_name
    github.event.commits.*.message
    github.event.head_commit.message
    github.event.head_commit.author.email
    github.event.head_commit.author.name
    github.event.commits.*.author.email
    github.event.commits.*.author.name
    github.event.pull_request.head.ref
    github.event.pull_request.head.label
    github.event.pull_request.head.repo.default_branch
    github.head_ref

For example, following action has a comand injection vulnerbility.

```yaml
name: test
on: [pull_request]
permissions:
  contents: read
jobs:
  test:
    name: "Test"
    runs-on: ubuntu-18.04
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - run: echo "${{ github.event.pull_request.title }}"
```

If attacker create malcious PR title like `";cat .git/config|base64"`, GitHub Actions will execulte the title as command.

📝 `actions/checkout` set `GITHUB_TOKEN` in `.git/config` for checkouting repo.

We can mitigate the vulnerbility of action by usging `env:` as parameter.

```yaml
name: test
on: [pull_request]
permissions:
  contents: read
jobs:
  test:
    name: "Test"
    runs-on: ubuntu-18.04
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - run: echo "${PR_TITLE}"
        env:
          PR_TITLE: "${{ github.event.pull_request.title }}"
```

If actionlint check the Untrusted input in `run:`, it make us safer.

## Related

- [GitHub Capture the Flag results | The GitHub Blog](https://github.blog/2021-03-22-github-ctf-results/)
- [Security hardening for GitHub Actions - GitHub Docs](https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#understanding-the-risk-of-script-injections)
- [PyPIにおける潜在的な任意コード実行 - RyotaK's Blog](https://blog.ryotak.me/post/pypi-potential-remote-code-execution/)

GitHub CodeQL has implemented related functions.

- [codeql/ExpressionInjection.ql at main · github/codeql](https://github.com/github/codeql/blob/main/javascript/ql/src/experimental/Security/CWE-094/ExpressionInjection.ql)
- [codeql/UntrustedCheckout.ql at main · github/codeql](https://github.com/github/codeql/blob/main/javascript/ql/src/experimental/Security/CWE-094/UntrustedCheckout.ql)

**Comment by rhysd** [comment]
Thank you for your suggestion. It also makes sense to me. Let me consider this on next minor version up.

**Comment by rhysd** [comment]
I've implemented this on [untrusted-input branch](https://github.com/rhysd/actionlint/tree/untrusted-input). Remaining task is adding tests and documents. This feature is being included in next minor version up v1.6.0.

**Comment by rhysd** [comment]
This feature was released at [v1.6.0](https://github.com/rhysd/actionlint/releases/tag/v1.6.0).

**Comment by azu** [comment]
Thanks!  Awesome!
